website/blog/2021-11-14-Presto-Cluster-Memory-Tuning.md

```diff
@@ -0,0 +1,70 @@
+
+Therefore, in order to prevent this kind of starvation, a space must be reserved for a large memory query to run. The reserved space is equal to the maximum memory allowed by the query. Every second, Presto picks out a query with the largest memory usage and allows it to use the reserved pool to avoid that there is always no available memory for the query to run.
+
+##Memory Tuning
+
+Presto has a couple of system configs on memory for the stability of the entire cluster:
+* `query.max-memory`: The maximum user memory allowed by a single query in the entire cluster
+* `query.max-memory-per-node`: The maximum user memory allowed by a query on a single worker
+* `query.max-total-memory`: The maximum (user + system) memory allowed by a single query in the entire cluster
+* `query.max-total-memory-per-node`: The maximum (user + system) memory allowed by a query on a single worker. If not set, by default it's 1/3 of heap size
+
+When these thresholds are exceeded, your query will be terminated with an insufficient memory error. Presto also has another memory related self-protection mechanism. When the general pool memory is used up, some operators will be set to a blocked state, allowing the operator to suspend for a while, thereby avoiding the entire system from crashing.
+
+Now for a cluster configured for concurrent queries, there are a few factors that should be considered when setting memory in a Presto Worker:
+
+* TotalMemory: Total amount of distributed memory a query may use
+* GeneralPool: User and system memory allocations may use
+* ReservedPool: reserved memory pool may be disabled by configuring `experimental.reserved-pool-enabled`.
+* Other: Operating system, some other processes in the system, and some off-heap memory, some operations in calculations require off-heap memory, such as some decompression algorithms
+* `memory.heap-headroom-per-node`: There are some memory allocations that cannot be counted by the Presto framework, such as some memory allocations in the third party library
+* `query.hash-partition-count`: The number of hash partitions in distributed joins and aggregations
+* `query.max-memory`: total memory allowed for a single query. When the user memory allocation of a query across all workers hits this limit, the query will be killed.
+* `query.max-total-memory-per-node`: Limits the total amount of memory a query may use on a single node
+
+Below is the relationship between these configs:
+
+* GeneralPool = TotalMemory-ReservedPool-`memory.heap-headroom-per-node`-Other
+* ReservedPool = `query.max-total-memory-per-node` (True only when user memory pool is exhausted. Reserved pool is as large as `query.max-total-memory-per-node` as revocable memory is kicked in to prevent cluster concurrency reduction. In general case, reserved pool is not used)
+* `query.max-memory` = `query.hash-partition-count` * `query.max-total-memory-per-node` / hash skew factor （where the skew factor is how much skew it can be tolerated)
```

Comment:
this one is a bit confusing to me be cause the other are factual descriptions about the relationship between, whereas this one is more guidance about how you ought to set query.max-memory.